pkgs/tools/nix/nixos-render-docs/src/tests/test_redirects.py

```diff
@@ -0,0 +1,228 @@
+            sources={
+                "foo.md": "# Foo {#foo}",
+                "bar.md": "# Bar {#bar}"
+            },
+            raw_redirects={"foo": ["foo.html"], "bar": ["foo.html"]},
+        )
+        self.assert_redirect_error({"identifiers_missing_current_outpath": ["bar"]}, intermediate)
+
+        after = self.setup_test(
+            sources={
+                "foo.md": "# Foo {#foo}",
+                "bar.md": "# Bar {#bar}"
+            },
+            raw_redirects={"foo": ["foo.html"], "bar": ["bar.html", "foo.html#bar"]},
+        )
+        self.run_test(after)
+
+    def test_non_leaf_identifier_moved_to_different_file(self):
+        """Test moving a non-leaf identifier to a different output path."""
+        before = self.setup_test(
+            sources={"foo.md": "# Foo {#foo}\n## Bar {#bar}\n### Baz {#baz}"},
+            raw_redirects={"foo": ["foo.html"], "bar": ["foo.html"], "baz": ["foo.html"]},
+        )
+        self.run_test(before)
+
+        intermediate = self.setup_test(
+            sources={
+                "foo.md": "# Foo {#foo}",
+                "bar.md": "# Bar {#bar}\n## Baz {#baz}"
+            },
+            raw_redirects={"foo": ["foo.html"], "bar": ["foo.html"], "baz": ["foo.html"]},
+        )
+        self.assert_redirect_error({"identifiers_missing_current_outpath": ["bar", "baz"]}, intermediate)
+
+        after = self.setup_test(
+            sources={
+                "foo.md": "# Foo {#foo}",
+                "bar.md": "# Bar {#bar}\n## Baz {#baz}"
+            },
+            raw_redirects={"foo": ["foo.html"], "bar": ["bar.html", "foo.html#bar"], "baz": ["bar.html", "foo.html#baz"]},
+        )
+        self.run_test(after)
+
+    def test_conflicting_anchors(self):
+        """Test for conflicting anchors."""
+        md = self.setup_test(
+            sources={"foo.md": "# Foo {#foo}\n## Bar {#bar}"},
+            raw_redirects={
+                "foo": ["foo.html", "foo.html#bar"],
+                "bar": ["foo.html"],
+            }
+        )
+        self.assert_redirect_error({"conflicting_anchors": ["bar"]}, md)
+
+    def test_divergent_redirect(self):
+        """Test for divergent redirects."""
+        md = self.setup_test(
+            sources={
+                "foo.md": "# Foo {#foo}",
+                "bar.md": "# Bar {#bar}"
+            },
+            raw_redirects={
+                "foo": ["foo.html", "old-foo.html"],
+                "bar": ["bar.html", "old-foo.html"]
+            }
+        )
+        self.assert_redirect_error({"divergent_redirects": ["old-foo.html"]}, md)
+
+    def test_client_path_with_server_redirect(self):
+        """Test for client paths with server redirects."""
+        md = self.setup_test(
+            sources={"foo.md": "# Foo {#foo}"},
+            raw_redirects={"foo": ["foo.html", "bar.html", "bar.html#foo"]}
+        )
+        self.assert_redirect_error({"client_paths_with_server_redirects": ["bar.html#foo"]}, md)
```

Comment:
I wonder how this an pass, because [here](https://github.com/NixOS/nixpkgs/pull/353513/files#diff-9bdf5c28b6cdf553d0d4e6735b57327691c4ad1bb8306e036c2b213fc460abe8R176) we're assigning dict fields

Reply:
Unsure, but I remember the unit test logs showing a list, not a dict -- again, not sure how that's possible; but I'll fix the types.